### PR TITLE
Only set DT_IMAGE_AUTO_PRESETS_APPLIED while in gui mode. Fixing #4586

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
The reporting issue is misleading, when watching the video the sequence is:
1. develop an image
2. discard history
3. export the image
4. reimport the image

The point is, after discarding the flag is cleared, export sets the flag but leaves no history and reimporting does **not** apply presets ... because the flag is already set.

How to fix? The auto apply should only be done while in gui.

### CODE REVIEW

**Comment by TurboGit** [comment]
Frankly to me this is not a bug. As I said to OP in the first issue I see nothing wrong.

You clear the history and export. Then the XMP is "frozen" as-is. If you have exported the picture without any history while would re-importing it should auto apply the presets?

**Comment by Grizzly1701** [comment]
@TurboGit in my opinion if there are presets that get auto applied by discarding history it should save the steps made by the presets in the xmp file. When reimporting it should just reopen the xmp file (and not change it/deleting the steps of the presets) which would mean that the steps automatically made by the presets would remain applied.

**Comment by parafin** [comment]
Discarding history should reset image to the state of freshly imported image, that means applying auto-applied presets, base curve included if it is enabled. This is an actual bug, no doubt about it.

**Comment by jenshannoschwalm** [comment]
@Grizzly1701 that's not the point. Discarding history **removes** the history and does not set it to something like default.

@Turbogit This is what i concider to be a bug.

If i have no history and export i would either expect
1. apply presets while exporting **and** keep them in history 
2. do not apply presets and wait for them to be applied when developing in darkroom.

Both ways seem to be ok for me. But here we apply presets while exporting, don't keep the history **but** have the flag switched on.

**Comment by jenshannoschwalm** [comment]
@parafin that's not exactly how it works. If i discard history dt removes history data, cleans up the xmp  and marks the image as not-having-applied-presets yet.

**Comment by parafin** [comment]
I'm talking about from the point-of-view of a user, not how it works internally. If user discards history, he should get the same state as if he just imported the image, whether he exports it of opens in darkroom.

**Comment by jenshannoschwalm** [comment]
@parafin yeah - sure - i misunderstood. That means you opt for 1. ?

**Comment by parafin** [comment]
I guess so, yes, it makes more sense.

**Comment by jenshannoschwalm** [comment]
ok and understood. That would require a different pr (this one chose 2. because that was what i expected)

let's see what @Turbogit and maybe other people think and i might close this and come back later ...

**Comment by TurboGit** [comment]
You can't apply preset while exporting as this means that you cannot export a plain RAW without any module activated. This is needed for creating noise profile IIRC and also to create a preset for the color lut module.

**Comment by TurboGit** [comment]
Oh, and I also export RAW without any presets to show the original picture and the one after development.

**Comment by parafin** [comment]
You can go into darkroom and create any history you like, even without any iops, but discard history button is not for that.

**Comment by TurboGit** [comment]
> You can go into darkroom and create any history you like, even without any iops, but discard history button is not for that.

I see and I agree for discard history. Yet if you have many images to export without any modules activated you'll have to go into darkroom one by one. If you want to create a noise profile out of 25 images it will be very painful. I suppose one can copy/paste deactivated modules, not impossible but still painful compared to what we have today.

I won't block this as I don't have string opinion so yes probably 1 is the way to go.

**Comment by parafin** [comment]
You can create an xmp with desired (empty) history and use with dt-cli or import in GUI

**Comment by jenshannoschwalm** [comment]
Ok. I see agreement on 1.

Will close this for now and look for a proper fix a.s.a.p.

**Comment by Grizzly1701** [comment]
> @Grizzly1701 that's not the point. Discarding history removes the history and does not set it to something like default.

Ok I understand your statement: Discard history takes away everything created by darktable (including presets). 
Why does the exported jpg file then look like it has the base curve? Why does the history stack in darkroom say "base curve" after using "discard history" and going into darkroom? Both give the impression of darktable applying the presets when using "discard history".

Either it shouldn't apply base curve when using "discard history" at all (means it shouldn't show base curve in darkroom before exporting and not putting the base curve onto the jpg) or it should apply it and write it into the xmp file (and then of course remain applied after reimporting the folder because it would just read the xmp file (as it does now too doesn't it?))

To sum up things I would of course agree more with solution 1 but I would not apply the presets while exporting. Wouldn't this mean if I go to darktable after "discard history" in lighttable it wouldn't show my presets in the history stack but then apply the presets on the jpgs/exported images (exported images would look different than the image that showed up in darkroom)? Correct me if I misunderstood your idea.

**Comment by jenshannoschwalm** [comment]
> Why does the exported jpg file then look like it has the base curve?

Because whenever an image is opened (whether in darkroom mode or lua, export or cli) it will apply all defaults if there no history. So your exported image is ok :-)

Internally there is wrong housekeeping about a flag in dt database and the history and xmp files. THIS is the bug that needs to be fixed.

It took me a while to understand that you clicked on discard history. This is something pretty unusual in the workflow so i didn't expect that. Normally it's only done if your image development is a complete mess and you want to start from the beginning.

BTW get yourself used to the history stack on the left side! See what happens while selecting earlier steps or compress history.

**Comment by Grizzly1701** [comment]
> Because whenever an image is opened (whether in darkroom mode or lua, export or cli) it will apply all defaults if there no history. So your exported image is ok :-)

It does not apply presets when opening the RAW file (double clicking in lighttable --> opening in darkroom) the second time though (where the history is discarded): https://youtu.be/GFaFrRnF9lQ?t=33 0:33 in my video.

**Comment by jenshannoschwalm** [comment]
Just wait for the next & correct pr and you will see everything is right :-)

**Comment by Grizzly1701** [comment]
Sure. I think the same :) 

Just wanted to make sure you understand me right.

Thank you for your time and efforts!

**Comment by jenshannoschwalm** [comment]
Sometimes it's difficult to understand what **exactly** is the technical bug but I think I got it 🤔
